radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
 #[allow(dead_code)]
 mod types;
 
-/// Inserts a new set of nodes at version `current_version` + 1 into the tree
-/// persisted within the given `store`.
-/// This inserts a new leaf node for each given "change", together with an
-/// entire new "parent chain" leading from that leaf to a new root node (common
-/// for all of them).
-/// Each change may either create/update a substate's value (denoted by
-/// `Some(hash(scrypto_encode(value)))`), or delete a substate (denoted by
-/// `None`).
-/// All nodes that became stale precisely due to this (i.e. not any previous)
-/// operation will be reported before the function returns (see
-/// `WriteableTreeStore::record_stale_node`).
-/// Returns the hash of the newly-created root (i.e. representing state at
-/// version `current_version` + 1).
+/// A change of value associated with some ID.
+/// External API only uses it for hashes of substates (see `SubstateHashChange`), but internally we
+/// split into "change of ReNode layer's leaf" and "change of substate offset's value in the nested
+/// tree".
+pub struct IdChange<I, V> {
+    /// ID.
+    id: I,
+    /// A value after change, or `None` if this change denotes a delete.
+    changed: Option<V>,
+}
+
+impl<I, V> IdChange<I, V> {
+    pub fn new(id: I, changed: Option<V>) -> Self {
+        Self { id, changed }
+    }
+}
+
+/// A top-level `IdChange`, representing an actual change of a specific substate's hashed value.
+pub type SubstateHashChange = IdChange<SubstateId, Hash>;
+
+/// Inserts a new set of nodes at version `current_version` + 1 into the "nested JMT" persisted
+/// within the given `store`.
+/// In a traditional JMT, this inserts a new leaf node for each given "change", together with an
+/// entire new "parent chain" leading from that leaf to a new root node (common for all of them).
+/// In our instantiation of the JMT, we first update all nested per-`ReNodeModule` trees (i.e. of
+/// each {`RENodeId`, `NodeModuleId`} pair encountered in the `changes`), and then we update the
+/// single upper-layer tree (representing all `ReNodeModule`).
+/// All nodes that became stale precisely due to this (i.e. not any previous) operation will be
+/// reported before the function returns (see `WriteableTreeStore::record_stale_node`).
+/// Returns the hash of the newly-created root (i.e. representing state at version
+/// `current_version` + 1).
 ///
 /// # Panics
-/// Panics if a root node for `current_version` does not exist. The caller
-/// should use `None` to denote an empty, initial state of the tree (i.e.
-/// inserting at version 1).
+/// Panics if a root node for `current_version` does not exist. The caller should use `None` to
```

Comment:
I think the API would be nicer if we don't pass `current_version: None`, and instead there was a separate method to create the pre-genesis tree at state version 0?

Then we pass `current_version: 0` in here for genesis?

Reply:
hmm, not that easy... a "pre-genesis tree" is literally "empty TreeStore" (not `NullNode`, nothing).

What you ask for is to "treat version=0 as pre-genesis".
This makes sense for the ReNode layer (because versions only ever advance there by 1), but please note it is no longer true for SubstateOffset layer (some substate may start at v=8, and it doesn't say anything about what previous version exists, if any). To me `None` is a very good indication of "no previous version exists". `0` would be more of a "hacky marker/sentinel" for that. Also, the original opensource authors use `None`, if that means anything (yea their API is not an ideal example, but still).

Reply:
I agree it makes sense for the subtrees 👍. I'm happy to keep it for the root tree for ease/consistency, it just gave me a slight wtf before I read the Rust Docs :)